File: sdks/csharp/src/NovaBilling/Types/PaymentMethodResponse.cs
using System.Text.Json;
using System.Text.Json.Serialization;
using NovaBilling.Core;

namespace NovaBilling;

[Serializable]
public record PaymentMethodResponse : IJsonOnDeserialized
{
    [JsonExtensionData]
    private readonly IDictionary<string, JsonElement> _extensionData =
        new Dictionary<string, JsonElement>();

    [JsonPropertyName("id")]
    public required string Id { get; set; }

    [JsonPropertyName("customerId")]
    public required string CustomerId { get; set; }

    [JsonPropertyName("provider")]
    public required string Provider { get; set; }

    [JsonPropertyName("type")]
    public required string Type { get; set; }

    [JsonPropertyName("tokenId")]
    public required string TokenId { get; set; }

    [JsonPropertyName("isDefault")]
    public required bool IsDefault { get; set; }

    [JsonPropertyName("last4")]
    public string? Last4 { get; set; }

    [JsonPropertyName("brand")]
    public string? Brand { get; set; }

    [JsonPropertyName("expMonth")]
    public double? ExpMonth { get; set; }

    [JsonPropertyName("expYear")]
    public double? ExpYear { get; set; }

    [JsonPropertyName("cardholderName")]
    public string? CardholderName { get; set; }

    [JsonPropertyName("country")]
    public string? Country { get; set; }

    [JsonPropertyName("createdAt")]
    public required DateTime CreatedAt { get; set; }

    [JsonPropertyName("updatedAt")]
    public required DateTime UpdatedAt { get; set; }

    [JsonIgnore]
    public ReadOnlyAdditionalProperties AdditionalProperties { get; private set; } = new();

    void IJsonOnDeserialized.OnDeserialized() =>
        AdditionalProperties.CopyFromExtensionData(_extensionData);

    /// <inheritdoc />
    public override string ToString()
    {
        return JsonUtils.Serialize(this);
    }
}

File: sdks/ruby/lib/novabilling/types/payment_method_response.rb
# frozen_string_literal: true

module Novabilling
  module Types
    class PaymentMethodResponse < Internal::Types::Model
      field :id, -> { String }, optional: false, nullable: false
      field :customer_id, -> { String }, optional: false, nullable: false, api_name: "customerId"
      field :provider, -> { String }, optional: false, nullable: false
      field :type, -> { String }, optional: false, nullable: false
      field :token_id, -> { String }, optional: false, nullable: false, api_name: "tokenId"
      field :is_default, -> { Internal::Types::Boolean }, optional: false, nullable: false, api_name: "isDefault"
      field :last4, -> { String }, optional: true, nullable: false
      field :brand, -> { String }, optional: true, nullable: false
      field :exp_month, -> { Integer }, optional: true, nullable: false, api_name: "expMonth"
      field :exp_year, -> { Integer }, optional: true, nullable: false, api_name: "expYear"
      field :cardholder_name, -> { String }, optional: true, nullable: false, api_name: "cardholderName"
      field :country, -> { String }, optional: true, nullable: false
      field :created_at, -> { String }, optional: false, nullable: false, api_name: "createdAt"
      field :updated_at, -> { String }, optional: false, nullable: false, api_name: "updatedAt"
    end
  end
end
